packages/langium-cli/src/generator/highlighting/monarch-generator.ts

```diff
@@ -0,0 +1,476 @@
+        return fmt('{ regex: ' + rulePatt + ', action: ' + prettyPrintAction(ruleOrState.action) + ' },');
+    } else {
+        // include another state by name, implicitly includes all of its contents
+        return fmt(`{ include: '@${ruleOrState.name}' },`);
+    }
+}
+
+/**
+ * Pretty prints the action of a Rule
+ * @param action Action to print. Can have several keywords to control what the state machine should do next.
+ * @returns Action in concrete form
+ */
+function prettyPrintAction(action: Action | Case[]): string {
+    if(!Array.isArray(action)) {
+        // plain action
+        return JSON.stringify(action);
+    } else {
+        // array of cases, each with an action
+        const prettyCases: string = action.map(c => `'${c.guard}': ` + prettyPrintAction(c.action)).join(', ');
+        return '{ cases: { ' + prettyCases + ' }}';
+    }
+}
+
+/**
+ * Convert a deafult Langium token names to a monarch one
+ * @param name Token name to convert
+ * @returns Returns the equivalent monarch name, or the original token name
+ */
+function getMonarchTokenName(name: string): string {
+    if(name === 'WS') {
+        return 'white';
+    } else if (name === 'ML_COMMENT' || name === 'SL_COMMENT') {
+        return 'comment';
+    } else if (name === 'STRING') {
+        return 'string';
+    } else if (name === 'INT') {
+        return 'number';
+    } else if (name === 'BIGINT') {
+        return 'number.float';
+    } else if (name === 'ID') {
+        return 'identifier';
+    } else {
+        // fallback to the original name
+        return name;
+    }
+}
+
+/**
+ * Gets whitespace rules from the langium grammar. Includes starting comment sequence
+ * @param grammar Langium grammar to extract whitespace rules from
+ * @returns Array of Monarch whitespace rules
+ */
+function getWhitespaceRules(grammar: langium.Grammar): Rule[] {
+    const rules: Rule[] = [];
+    for(const rule of grammar.rules) {
+        if(isTerminalRule(rule) && (isCommentTerminal(rule) || rule.name === 'WS') && isRegexToken(rule.definition)) {
+            const tokenName = getMonarchTokenName(rule.name);
+            const part = getTerminalParts(terminalRegex(rule))[0];
+            if(part.start !== '' && part.end !== '' && tokenName === 'comment') {
+                // state-based rule, only add push to jump into it
+                rules.push({
+                    regex: part.start.replace('/', '\\/'),
+                    action: { token: tokenName, next: '@' + tokenName }
+                });
+
+            } else {
+                // single regex rule
+                rules.push({
+                    regex: rule.definition.regex,
+                    action: {token: getMonarchTokenName(rule.name) }
+                });
+            }
+        }
+    }
+    return rules;
+}
+
+/**
+ * Gets comment state rules from the Langium grammar. Accounts for nested multi-line comments.
```

Comment:
We might not actually want to account for nested multi-line rules, since Chevrotain can't parse those.

Reply:
Initially I was using the original ML_COMMENT regex, but found some issues with the nested case. If I'm not mistaken, isn't this similar to the approach we're using in TextMate (sans the stack) by specifying our start/stop sequence? We could perhaps still mirror that solution, but w/out the stack?

Reply:
Yeah. I'm not sure whether textmate uses a stack. Anyway, we should always try that the monarch/textmate highlighting is in sync with what's actually getting parsed.

Reply:
Seems like [TextMate supports nesting of patterns for the begin/end rule keys](https://macromates.com/manual/en/language_grammars#rule_keys). Although to keep it the same as what our current TM grammar generation is, I've settled just for simple start -> end match, with no nesting for comments.